tpke/src/context.rs

```diff
@@ -22,3 +29,15 @@ pub struct PrivateDecryptionContext<E: PairingEngine> {
     pub scalar_bits: usize,
     pub window_size: usize,
 }
+
+#[derive(Clone, Debug)]
+pub struct PrivateDecryptionContextSimple<E: PairingEngine> {
+    pub index: usize,
+    pub b: E::Fr,
+    pub b_inv: E::Fr,
+    pub private_key_share: PrivateKeyShare<E>,
+    pub public_decryption_contexts: Vec<PublicDecryptionContextSimple<E>>,
+    pub g: E::G1Affine,
+    pub g_inv: E::G1Prepared,
+    pub h_inv: E::G2Prepared,
```

Comment:
I agree, we could refactor them into `SetupParams`. I was also considering removing the fast variant altogether (per my previous comment). That would leave us with no duplication.

On another note, Rust uses composition over inheritance. So instead of 
```
struct Parent {
  a: u32,
  b: u32,
}

struct Child extends Parent {
  c: u32,
}
```
we have
```
struct Child {
  parent: Parent,
  c: u32,
}
```
I think in this case composition still works just fine, I just thought I could mention this.

Reply:
Composition sounds great here